### ISSUE TITLE
Auto Rerun Pull Request Builds when Target Branch has a new Commit?

### ISSUE THREAD

**Issue by drdamour** [travis-ci/travis-ci#1620]
Suppose a pull request exists that is merging **_commit1**_ and **_commit2**_ from branch **_Source**_ into branch **_Target**_

Travis has already run the tests for the pull request and found commit1 & commit2 to be passing within the pull request.

Now someone makes **_commit3**_ to **_Target**_ that does NOT cause a merge conflict in the pull request.

This may fundamentally affect the pull request even without causing a merge conflict.  If you were to checkout the PR commit and run the test suite it may now fail.  Or a PR that had previously been marked as a failure may now actually pass.

This appears to be quite common in our gitflow (maybe our gitflow is just bad).  As such i was surprised Travis didn't do anything about this.

What i was expecting was that Travis would determine that **_commit3**_ was applied to **_Target**_ branch it would go and look for any pull requests that are merging into **_Target**_ that still can be auto merged.  For each found PR travis would automatically rerun the builds for that PR putting them in yellow status until their overall merge status is determined.

Is there a reason for it not to do this?  Would it cause undesired behaviour? or is it just a feature that hasn't been thought of or implemented yet?  Sorry if it's a duplicate.

**Comment by dregad** [comment]
> you can circumvent this shortcoming on a per pull request basis by simply restarting the last build for the PR

@rkh I just tried that, but it only seems to restart the build at the original commit instead of using the PR's source branch head into consideration. Here is what I did:

- Go to PR tab https://travis-ci.org/mantisbt/mantisbt/pull_requests
- Open the PR's build [#2318](https://travis-ci.org/mantisbt/mantisbt/builds/101091492)
- Click the *Restart build* button in top right

Did I miss something ?

**Comment by percyhanna** [comment]
GitHub's new "protected branches" allows you to force PRs to be up-to-date with the target branch before being merged. However, this will require manual intervention on each PR owner every time `master` is updated, which seems more frustrating than it's worth. Having this as an optional feature would be preferred, so you can decide whether to flood the build queue or not. For a smaller team with 5 build slots, it wouldn't be terrible given the convenience of ensuring `master` never breaks accidentally.

**Comment by tibdex** [comment]
> GitHub's new "protected branches" allows you to force PRs to be up-to-date with the target branch before being merged. However, this will require manual intervention on each PR owner every time `master` is updated, which seems more frustrating than it's worth.

I like the option to force PRs to be up-to-date, it protects the repo from ["semantic conflicts"](https://bors.tech/essay/2017/02/02/pitch/). But I agree that the manual intervention involved with rebasing the PR every time `master` is updated is annoying. That's why I developed this GitHub App: https://github.com/tibdex/autorebase. I hope you find it useful 😉.

**Comment by aryairani** [comment]
It would be really valuable from our perspective to be able to have CI _automatically_ detect prospective build failures before they're merged into `master` rather than afterwards.

Even if there's agreement that we shouldn't launch 20 new builds in each commit, it's important to visually invalidate the previous ✅success and provide an easy way of refreshing the result.  

Or better yet might be to debounce or something so that the results could automatically refresh eventually, so you aren't guaranteed to always have to stop and wait on a manual build before merging an older PR.

**Comment by Alizter** [comment]
Could we have some sort of indication that the build is stale so that we know we need to restart it?

**Comment by doivosevic** [comment]
Hey, is there any chance this issue can be reopened? We have encountered this issue and I'm surprised more people aren't asking for this. Without this you cannot guarantee the develop branch is passing the CI checks. Is there a way to handle this right now without including 3rd party libraries nor manually turning branches into protected?
